server/comms/server.go

```diff
@@ -158,30 +177,100 @@ type RPCConfig struct {
+	// confident it's not going to interfere with init/redeem or others.
+	// if !rl.cumulative.Allow() {
+	// 	return false
+	// }
+	limiter := rl.routes[route]
+	if limiter == nil {
+		return true // free
+	}
+	return rl.cumulative.Allow() && limiter.Allow()
+}
+
+// newRouteLimiter creates a route-based rate limiter. It should be applied to
+// all connections from a given IP address.
+func newRouteLimiter() *routeLimiter {
+	// Some routes share a limiter to aggregate request stats:
+	statusLimiter := rate.NewLimiter(wsRateStatus, wsBurstStatus)
+	orderLimiter := rate.NewLimiter(wsRateOrder, wsBurstOrder)
+	infoLimiter := rate.NewLimiter(wsRateInfo, wsBurstInfo)
+	return &routeLimiter{
+		cumulative: rate.NewLimiter(wsRateTotal, wsBurstTotal),
+		routes: map[string]allower{
+			// Meter the 'register' route the most.
+			msgjson.RegisterRoute: rate.NewLimiter(wsRateRegister, wsBurstRegister),
+			// Status checking of matches and orders
+			msgjson.MatchStatusRoute: statusLimiter,
+			msgjson.OrderStatusRoute: statusLimiter,
+			// Order submission
+			msgjson.LimitRoute:  orderLimiter,
+			msgjson.MarketRoute: orderLimiter,
+			msgjson.CancelRoute: orderLimiter,
+			// Order book subscription with full snapshot
+			msgjson.OrderBookRoute: rate.NewLimiter(wsRateBook, wsBurstBook),
+			// Config and fee rate
+			msgjson.FeeRateRoute: infoLimiter,
+			msgjson.ConfigRoute:  infoLimiter,
+		},
+	}
+}
+
+// ipWsLimiter facilitates connection counting for a source IP address to
+// aggregate requests stats by a single rate limiter.
+type ipWsLimiter struct {
+	conns   int64
+	cleaner *time.Timer // when conns drops to zero, set a cleanup timer
+	*routeLimiter
+}
+
 // Server is a low-level communications hub. It supports websocket clients
 // and an HTTP API.
 type Server struct {
 	// One listener for each address specified at (RPCConfig).ListenAddrs.
 	listeners []net.Listener
-	// Protect the client map, which maps the (link).id to the client itself.
+
+	// The client map indexes each wsLink by its id.
 	clientMtx sync.RWMutex
 	clients   map[uint64]*wsLink
-	// A simple counter for generating unique client IDs. The counter is also
-	// protected by the clientMtx.
-	counter uint64
+	counter   uint64 // for generating unique client IDs
+
+	// wsLimiters manages per-IP per-route websocket connection request rate
+	// limiters that are not subject to server-wide rate limits or affected by
+	// disabling of the data API (Server.dataEnabled).
+	wsLimiterMtx sync.Mutex // the map and the fields of each limiter
+	wsLimiters   map[dex.IPKey]*ipWsLimiter
+	v6Prefixes   map[dex.IPKey]int // just debugging presently
+
 	// The quarantine map maps IP addresses to a time in which the quarantine will
 	// be lifted.
-	banMtx      sync.RWMutex
-	quarantine  map[dex.IPKey]time.Time
-	dataEnabled uint32
+	banMtx     sync.RWMutex
+	quarantine map[dex.IPKey]time.Time
+
+	dataEnabled uint32 // atomic
```

Comment:
Best to have atomics at the top of the struct? Something about 32 bit systems.

Reply:
That particular atomic bug affects 64-bit data on 32-bit systems.  But it still may be good practice to put them first.  Will think about that.

Reply:
Ok, 64 bit atomics on old 32 bit processors. I will remember this from now.